x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/audit/logfile/LoggingAuditTrail.java

```diff
@@ -762,29 +768,40 @@ LogEntryBuilder with(String key, String[] values) {
             return this;
         }
 
+        LogEntryBuilder with(Map<String, Object> map) {
+            for (Entry<String, Object> entry : map.entrySet()) {
+                Object value = entry.getValue();
+                if (value.getClass().isArray()) {
+                    logEntry.with(entry.getKey(), toQuotedJsonArray((Object[]) value));
+                } else {
+                    logEntry.with(entry.getKey(), value);
+                }
+            }
+            return this;
+        }
+
         StringMapMessage build() {
             return logEntry;
         }
 
-        String toQuotedJsonArray(String[] values) {
+        String toQuotedJsonArray(Object[] values) {
             assert values != null;
             final StringBuilder stringBuilder = new StringBuilder();
             final JsonStringEncoder jsonStringEncoder = JsonStringEncoder.getInstance();
             stringBuilder.append("[");
-            for (final String value : values) {
+            for (final Object value : values) {
                 if (value != null) {
                     if (stringBuilder.length() > 1) {
                         stringBuilder.append(",");
                     }
                     stringBuilder.append("\"");
-                    jsonStringEncoder.quoteAsString(value, stringBuilder);
+                    jsonStringEncoder.quoteAsString(value.toString(), stringBuilder);
```

Comment:
I just noticed this while review the backport:
This won't be null-safe (for example values = `new Object[] { null }`)
```suggestion
                    jsonStringEncoder.quoteAsString(String.valueof(value), stringBuilder);
```

Reply:
Grr, I just saw the `!= null` above. Ignore me.